File: Data Structures & Algorithms/climbing-stairs/submission-18.py
class Solution:
    def climbStairs(self, n: int) -> int:
        def rec(n, memo): 
            if n in memo: 
                return memo[n] 
            if n == 1: 
                return 1 
            if n == 2: 
                return 2 

            memo[n] = rec(n - 1, memo) + rec(n - 2, memo) 
            return memo[n] 

        return rec(n, {}) 

File: Data Structures & Algorithms/decode-ways/submission-18.js
class Solution {
    /**
     * @param {string} s
     * @return {number}
     */
    numDecodings(s) { 

        function rec(i, memo) { 
            if (i in memo) return memo[i]
            if (i === s.length) return 1 
            if (s[i] === '0') return 0 

            let res = 0 
            res = rec(i + 1, memo) 

            if (i + 1 < s.length && Number(s.slice(i, i + 2)) <= 26) {
                res += rec(i + 2, memo)
            }  

            memo[i] = res
            return res
        } 

        return rec(0, {})
    }
}
